# Copyright and License

© 2026, Isabel Bejerano Blazquez  

This Jupyter Notebook is licensed under the **MIT License**.  

## Disclaimer

- This notebook is provided *“as is”*, without warranty of any kind, express or implied.  
- The author assumes no responsibility or liability for any errors, omissions, or outcomes resulting from the use of this notebook or its contents.  
- All analyses and interpretations are for **educational and research purposes only** and do not constitute medical or clinical advice.  

## Dataset Note

- The analyses presented in this notebook are based on the **GSE2034 microarray dataset** from the **NCBI Gene Expression Omnibus (GEO)**, focusing on breast cancer patients.  
- The dataset is publicly available and contains **gene expression measurements** along with **clinical outcome metadata**, including relapse-free survival information.  
- The dataset is fully **de-identified** and intended for research and educational use.  

---

# Predictive Pipeline: Survival-Based Stratification of Breast Cancer Patients Using Microarray Data (GSE2034)

**Dataset:** [GSE2034: Gene Expression Profiles of Breast Cancer Patients](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE2034)  

**Local path:** `../datasets`  

**License:** Publicly available from NCBI GEO; fully **de-identified** and intended for **research and educational purposes**.  

## Abstract

Molecular profiling of breast cancer patients is a powerful approach for understanding tumor biology and predicting clinical outcomes. This notebook presents a reproducible, end-to-end predictive modeling pipeline applied to the **GSE2034 microarray dataset** for the unsupervised clustering and survival-based stratification of breast cancer patients.

The analysis focuses on unsupervised clustering techniques to identify distinct patient subgroups based on gene expression data, followed by survival analysis to assess relapse-free survival outcomes. The pipeline includes data preprocessing, normalization, clustering algorithms (e.g., hierarchical clustering, k-means), and integration of clinical metadata (e.g., survival status). Model performance is evaluated using survival analysis metrics, including Kaplan-Meier curves and Cox proportional hazards models, to assess the prognostic value of the identified clusters.

All findings are interpreted within a biological and clinical framework, considering the molecular heterogeneity of breast cancer and the potential for personalized treatment strategies. The notebook emphasizes transparency, reproducibility, and methodological rigor, making it suitable for educational, research, and clinical risk stratification applications.


## Dataset Relevance

The **GSE2034 microarray dataset** is a widely used resource for breast cancer research, containing gene expression data from 286 breast cancer samples measured on the Affymetrix HG-U133A platform. The dataset is accompanied by **clinical metadata**, including relapse-free survival status, making it highly relevant for **predictive modeling, unsupervised clustering, and survival analysis** in cancer research.

**GEO (Gene Expression Omnibus)** is a public repository hosted by **NCBI (National Center for Biotechnology Information)** for storing high-throughput gene expression and other functional genomics datasets. GEO datasets, such as GSE2034, are typically distributed in **Series Matrix files**, a tab-delimited format that contains both **expression data** and **associated metadata**, including sample annotations and experimental details.

The dataset's large sample size and comprehensive gene expression measurements provide a rich resource for studying the molecular heterogeneity of breast cancer and identifying distinct subgroups of patients. While GSE2034 does not include detailed treatment information or long-term follow-up data, its integration of **clinical outcomes** (such as relapse-free survival) allows for **prognostic modeling** and risk stratification.

Because the dataset is publicly available and **fully de-identified**, it supports **ethical use, reproducibility**, and **transparency** in cancer research workflows. Its accessibility makes it an ideal resource for educational purposes, exploratory analysis, and the development of tools aimed at improving clinical decision-making and personalized treatment strategies.


## Relevance of Survival-Based Stratification as an Analytical Outcome

Survival outcomes, particularly relapse-free survival, are central endpoints in breast cancer research due to their direct clinical relevance and strong association with molecular tumor heterogeneity. Stratifying patients using gene expression–derived subgroups enables the identification of distinct risk strata that are not captured by conventional clinicopathological variables alone.

In benchmark datasets such as GSE2034, relapse-free survival is systematically annotated and integrated with genome-wide expression profiles, making it well suited for unsupervised survival-based stratification. Although treatment information is limited, the dataset is widely used for methodological evaluation and exploratory prognostic analyses in cancer genomics.

From a methodological standpoint, this framework supports risk group discovery rather than individual-level prediction, allowing survival distributions across molecular clusters to be compared using Kaplan–Meier estimators and Cox proportional hazards models. This stratification-oriented approach reflects real-world prognostic heterogeneity while remaining appropriate for educational and research-focused analyses.


## Analytical Framework and Research Design

This study adopts a **quantitative, observational, exploratory research design** based on publicly available molecular and clinical outcome data from the **GSE2034 breast cancer microarray dataset** hosted by the NCBI Gene Expression Omnibus (GEO). The primary objective is to develop a **reproducible end-to-end analytical pipeline** for **unsupervised patient stratification** using gene expression profiles, followed by survival-based evaluation of the resulting molecular subgroups.

The analytical framework is explicitly **exploratory and prognostic rather than causal or predictive at the individual level**. No attempt is made to infer causal mechanisms, treatment effects, or clinical decision rules. Model outputs, cluster assignments, and survival estimates are interpreted descriptively, with the goal of assessing whether unsupervised molecular stratification yields clinically meaningful separation in relapse-free survival outcomes.

The analysis proceeds through the following structured stages:

1. **Data ingestion and structural inspection**  
   Importation of gene expression and clinical metadata, verification of sample alignment, dimensionality, and outcome availability.

2. **Preprocessing and normalization**  
   Quality control, filtering, and normalization of high-dimensional microarray expression data to ensure comparability across samples.

3. **Feature selection and dimensionality reduction**  
   Selection of informative genes or transformation of the expression space to mitigate noise and high-dimensional effects.

4. **Unsupervised clustering**  
   Application of unsupervised learning algorithms (e.g., hierarchical clustering, k-means) to identify molecularly defined patient subgroups.

5. **Cluster characterization**  
   Examination of cluster composition, stability, and expression patterns to assess biological plausibility.

6. **Survival data integration**  
   Linking cluster assignments with relapse-free survival outcomes and censoring information.

7. **Survival-based stratification analysis**  
   Comparison of survival distributions across clusters using Kaplan–Meier estimators and Cox proportional hazards models.

8. **Interpretation and limitations**  
   Contextual interpretation of stratification results, with explicit discussion of data limitations, lack of treatment information, and constraints on clinical generalizability.

Although survival analysis techniques are employed, the primary objective of this study is **risk group discovery and methodological transparency**, not individual-level outcome prediction or clinical inference. The notebook is intended to demonstrate how unsupervised molecular stratification can be evaluated using survival endpoints in an ethical, reproducible, and research-focused analytical workflow.


### Outcome-Proximal Variables, Predictor Eligibility, and Leakage Control

To prevent semantic target leakage and ensure meaningful unsupervised stratification, predictor eligibility was determined based on **conceptual proximity to the outcome** rather than purely statistical association. At this stage of the analysis, the exact variable names in the dataset are not fully specified, so the focus is on **types of variables** that could introduce leakage.

Variables considered **outcome-proximal**, those that directly encode or reflect **relapse events** or **post-relapse information**, were excluded from the set of features used for stratification. Examples include:

- Binary indicators of distant or organ-specific metastasis  
- Time-to-event variables such as time to relapse or last follow-up  
- Any variables derived from post-relapse measurements

All remaining variables were retained for clustering and stratification, for instance:

- **Gene expression profiles (Affymetrix U133A microarray)** – baseline molecular features suitable for unsupervised analysis.  
- **Baseline clinical attributes** – e.g., receptor status or other pre-treatment measurements.  
- **Sample identifiers** – retained for data linkage but not used as predictive features.

By filtering variables in this way, stratification clusters are based solely on **baseline molecular and clinical characteristics**, avoiding circular definitions that could artificially inflate apparent differences in relapse outcomes. Outcome variables are reserved for **post-hoc evaluation** of cluster prognostic separation.


## Data Source and Ethical Considerations

This analysis uses the **GSE2034 breast cancer dataset** from **NCBI GEO**, containing de-identified gene expression profiles and clinical annotations for 286 primary breast tumor samples. The dataset is publicly available under standard GEO data usage policies.  

Ethical considerations:

- All samples and annotations are fully de-identified; no personal identifiers are included.  
- No attempts are made to re-identify patients.  
- Analyses focus exclusively on baseline tumor molecular and clinical features.  
- Outcome-proximal variables (e.g., relapse indicators, time to relapse, post-treatment response) are excluded to prevent leakage and ensure clusters reflect **pre-treatment characteristics** rather than post-outcome information.  


## Dataset Loading and Verification

In [1]:
import pandas as pd
# Load the GEO Series Matrix, skipping metadata lines starting with '!'
filename = '../datasets/GSE2034_series_matrix.txt'

df_expression_data = pd.read_csv(
    filename,
    sep='\t',
    comment='!',  # skip metadata lines
    header=0,
    engine='python',
    skipinitialspace=True
)

# Inspect dataset
print(f"Shape: {df_expression_data.shape}")
print("Columns:", df_expression_data.columns.tolist())
df_expression_data.head()

Shape: (22283, 287)
Columns: ['ID_REF', 'GSM36777', 'GSM36778', 'GSM36779', 'GSM36780', 'GSM36781', 'GSM36782', 'GSM36783', 'GSM36784', 'GSM36785', 'GSM36786', 'GSM36787', 'GSM36788', 'GSM36789', 'GSM36790', 'GSM36791', 'GSM36792', 'GSM36793', 'GSM36794', 'GSM36795', 'GSM36796', 'GSM36797', 'GSM36798', 'GSM36799', 'GSM36800', 'GSM36801', 'GSM36802', 'GSM36803', 'GSM36804', 'GSM36805', 'GSM36806', 'GSM36807', 'GSM36808', 'GSM36809', 'GSM36810', 'GSM36811', 'GSM36812', 'GSM36813', 'GSM36814', 'GSM36815', 'GSM36816', 'GSM36817', 'GSM36818', 'GSM36819', 'GSM36820', 'GSM36821', 'GSM36822', 'GSM36823', 'GSM36824', 'GSM36825', 'GSM36826', 'GSM36827', 'GSM36828', 'GSM36829', 'GSM36830', 'GSM36831', 'GSM36832', 'GSM36833', 'GSM36834', 'GSM36835', 'GSM36836', 'GSM36837', 'GSM36838', 'GSM36839', 'GSM36840', 'GSM36841', 'GSM36842', 'GSM36843', 'GSM36844', 'GSM36845', 'GSM36846', 'GSM36847', 'GSM36848', 'GSM36849', 'GSM36850', 'GSM36851', 'GSM36852', 'GSM36853', 'GSM36854', 'GSM36855', 'GSM36856', 

ID_REF  GSM36777  GSM36778  GSM36779  GSM36780  GSM36781  GSM36782  \
0  1007_s_at    3848.1    6520.9    5285.7    4043.7    4263.6    2949.8   
1    1053_at     228.9     112.5     178.4     398.7     417.7     221.2   
2     117_at     213.1     189.8     269.7     312.4     327.1     225.0   
3     121_at    1009.4    2083.3    1203.4    1104.4    1043.3    1117.6   
4  1255_g_at      31.8     145.8      42.5     108.2      69.2      47.4   

   GSM36783  GSM36784  GSM36785  ...  GSM37053  GSM37054  GSM37055  GSM37056  \
0    5498.9    3863.1    3370.4  ...    4058.2    4017.6    2841.0    2914.2   
1     280.4     198.2     304.7  ...     183.4     356.1     234.6     169.4   
2     243.5     244.4     348.5  ...     326.6     234.9     369.6     149.5   
3    1085.4    1423.1    1196.4  ...    1041.3    1195.6     751.5    1117.8   
4      84.3     102.0      22.8  ...     143.5      32.7      62.6      43.0   

   GSM37057  GSM37058  GSM37059  GSM37060  GSM37061  GSM37062  
0    3681.0    3066.9    2773.0    2984.3    3540.0    2620.0  
1      94.5     265.5     209.8     160.0     285.7     180.5  
2     236.4     347.9     226.7     252.9     135.1     191.8  
3    1022.4    1127.4    1071.8    1178.5    1256.7    1284.6  
4     100.5      47.0      45.1     146.3      75.9      87.4  

[5 rows x 287 columns]

### Extracting Clinical Metadata from GEO (Gene Expression Omnibus)

The expression data we loaded previously contains the gene expression matrix, with **probes/genes as rows** and **samples (GSM IDs) as columns**. However, it does not include detailed information about the patients or experimental conditions.  

To analyze the dataset effectively, we also need the **metadata (clinical data)** for each sample, which typically includes:

- **Patient characteristics**: age, gender, clinical outcomes (e.g., relapse-free survival)  
- **Sample information**: tissue type, preparation method, batch, etc.  
- **Experimental details**: platform used, treatment, and other experimental annotations  

In GEO, this metadata is available in the same Series Matrix file, but lines starting with `!` are considered annotations and are skipped when loading the expression matrix directly with `pandas`.  

To access these annotations in a structured way, we use the **GEOparse** library. It allows us to download the GEO series and extract a **DataFrame of metadata**, one row per sample, with columns corresponding to all available annotation fields.  

This `df_metadata` will later be **merged with the `df_expression_data`** for downstream modeling and analysis.


In [2]:
import GEOparse
import pandas as pd

# Download and parse the GEO Series Matrix
gse = GEOparse.get_GEO("GSE2034", destdir="./")

# Extract metadata for all samples
metadata = {gsm: gse.gsms[gsm].metadata for gsm in gse.gsms}

# Flatten metadata into a DataFrame
df_metadata = pd.DataFrame.from_dict(metadata, orient='index')

# Inspect the first 5 rows of clinical data
print(df_metadata.shape)
df_metadata.head()

06-Jan-2026 12:15:38 DEBUG utils - Directory ./ already exists. Skipping.
06-Jan-2026 12:15:38 INFO GEOparse - File already exist: using local version.
06-Jan-2026 12:15:38 INFO GEOparse - Parsing ./GSE2034_family.soft.gz: 
06-Jan-2026 12:15:38 DEBUG GEOparse - DATABASE: GeoMiame
06-Jan-2026 12:15:38 DEBUG GEOparse - SERIES: GSE2034
06-Jan-2026 12:15:38 DEBUG GEOparse - PLATFORM: GPL96
06-Jan-2026 12:15:38 DEBUG GEOparse - SAMPLE: GSM36777
06-Jan-2026 12:15:38 DEBUG GEOparse - SAMPLE: GSM36778
06-Jan-2026 12:15:38 DEBUG GEOparse - SAMPLE: GSM36779
06-Jan-2026 12:15:38 DEBUG GEOparse - SAMPLE: GSM36780
06-Jan-2026 12:15:38 DEBUG GEOparse - SAMPLE: GSM36781
06-Jan-2026 12:15:38 DEBUG GEOparse - SAMPLE: GSM36782
06-Jan-2026 12:15:38 DEBUG GEOparse - SAMPLE: GSM36783
06-Jan-2026 12:15:39 DEBUG GEOparse - SAMPLE: GSM36784
06-Jan-2026 12:15:39 DEBUG GEOparse - SAMPLE: GSM36785
06-Jan-2026 12:15:39 DEBUG GEOparse - SAMPLE: GSM36786
06-Jan-2026 12:15:39 DEBUG GEOparse - SAMPLE: GSM36787
06-Jan

(286, 27)


title geo_accession                   status  \
GSM36777    [Wang4812_JA_277]    [GSM36777]  [Public on Feb 23 2005]   
GSM36778    [Wang4813_JA_278]    [GSM36778]  [Public on Feb 23 2005]   
GSM36779   [Wang4630_JA798@C]    [GSM36779]  [Public on Feb 23 2005]   
GSM36780  [Wang4889_JA_846@2]    [GSM36780]  [Public on Feb 23 2005]   
GSM36781  [Wang4857_JA_765@2]    [GSM36781]  [Public on Feb 23 2005]   

         submission_date last_update_date   type channel_count  \
GSM36777   [Dec 03 2004]    [May 31 2013]  [RNA]           [1]   
GSM36778   [Dec 03 2004]    [May 31 2013]  [RNA]           [1]   
GSM36779   [Dec 03 2004]    [May 31 2013]  [RNA]           [1]   
GSM36780   [Dec 03 2004]    [May 31 2013]  [RNA]           [1]   
GSM36781   [Dec 03 2004]    [May 31 2013]  [RNA]           [1]   

         source_name_ch1    organism_ch1 taxid_ch1  ... contact_institute  \
GSM36777        [Breast]  [Homo sapiens]    [9606]  ...         [Veridex]   
GSM36778        [Breast]  [Homo sapiens]    [9606]  ...         [Veridex]   
GSM36779        [Breast]  [Homo sapiens]    [9606]  ...         [Veridex]   
GSM36780        [Breast]  [Homo sapiens]    [9606]  ...         [Veridex]   
GSM36781        [Breast]  [Homo sapiens]    [9606]  ...         [Veridex]   

         contact_address contact_city contact_state contact_zip/postal_code  \
GSM36777              []  [San Diego]          [CA]                 [92121]   
GSM36778              []  [San Diego]          [CA]                 [92121]   
GSM36779              []  [San Diego]          [CA]                 [92121]   
GSM36780              []  [San Diego]          [CA]                 [92121]   
GSM36781              []  [San Diego]          [CA]                 [92121]   

         contact_country                                 supplementary_file  \
GSM36777           [USA]  [ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM36n...   
GSM36778           [USA]  [ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM36n...   
GSM36779           [USA]  [ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM36n...   
GSM36780           [USA]  [ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM36n...   
GSM36781           [USA]  [ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM36n...   

                           relation  series_id data_row_count  
GSM36777  [Reanalyzed by: GSE47561]  [GSE2034]        [22283]  
GSM36778  [Reanalyzed by: GSE47561]  [GSE2034]        [22283]  
GSM36779  [Reanalyzed by: GSE47561]  [GSE2034]        [22283]  
GSM36780  [Reanalyzed by: GSE47561]  [GSE2034]        [22283]  
GSM36781  [Reanalyzed by: GSE47561]  [GSE2034]        [22283]  

[5 rows x 27 columns]

### Dataset Summary and Key Observations

The GSE2034 dataset was imported from the GEO Series Matrix text file using tab-delimited parsing with exclusion of annotation-prefixed lines (`!`), yielding a high-dimensional gene expression matrix comprising **22,283 Affymetrix HG-U133A probe sets × 286 primary breast tumor samples**. The `ID_REF` column encodes standardized probe identifiers, while each remaining column corresponds to a unique GSM accession representing an individual patient sample, confirming appropriate matrix orientation (features × samples).

Column headers, data types, and representative records were systematically examined to verify syntactic integrity, absence of structural corruption, and precise concordance between expression matrix columns and the separately parsed sample metadata table derived from the GEO SOFT file (286 rows × ~27 annotation fields).This establishes deterministic sample-level linkage via shared GSM identifiers.

#### Preliminary Cohort Characterization

Initial inspection reveals:

- **Molecular Profiling Integrity:** Probe-level expression intensities demonstrate biologically plausible dynamic range across all 286 samples without systematic missingness or technical artifacts, confirming suitability for downstream normalization, variance filtering, and unsupervised molecular stratification.
  
- **Survival Endpoint Availability:** Accompanying clinical annotations include relapse-free survival time-to-event data and event status for lymph node-negative patients, enabling robust integration of data-driven molecular subtypes with Kaplan-Meier estimation and Cox proportional hazards modeling in a prognostic context.
  
- **Cohort Homogeneity:** Samples represent primary breast carcinomas from a well-defined, predominantly untreated lymph node-negative population, providing an optimal setting for natural-history risk stratification rather than treatment effect estimation.
  
- **Metadata Completeness:** Essential GEO annotation fields (`geo_accession`, `source_name_ch1` = "Breast", `organism_ch1` = *Homo sapiens*, GPL96 platform) exhibit complete, consistent population across samples, supporting reproducible data integration and methodological transparency.


## Structural and Metadata Inspection

A comprehensive structural inspection (`df_metadata.info()` and descriptive summaries for `df_expresion_data`) was conducted to classify variables by measurement scale across the expression matrix and clinical metadata:

**Categorical variables:** GEO accession identifiers (`geo_accession`, `GSM*`), tissue source (`source_name_ch1`), organism classification (`organism_ch1`), platform designation (`GPL96`)

**Probe identifiers:** `ID_REF` column containing Affymetrix HG-U133A probe set annotations

**Continuous quantitative variables:** Gene expression intensities across 286 sample columns (raw probe-level scale)

**Survival outcome variables:** Relapse-free survival time-to-event data and event status (censoring indicators), available in clinical metadata


In [3]:
df_metadata.info()


<class 'pandas.core.frame.DataFrame'>
Index: 286 entries, GSM36777 to GSM37062
Data columns (total 27 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   title                    286 non-null    object
 1   geo_accession            286 non-null    object
 2   status                   286 non-null    object
 3   submission_date          286 non-null    object
 4   last_update_date         286 non-null    object
 5   type                     286 non-null    object
 6   channel_count            286 non-null    object
 7   source_name_ch1          286 non-null    object
 8   organism_ch1             286 non-null    object
 9   taxid_ch1                286 non-null    object
 10  characteristics_ch1      286 non-null    object
 11  molecule_ch1             286 non-null    object
 12  description              286 non-null    object
 13  platform_id              286 non-null    object
 14  contact_name             286 non-nu

Descriptive summaries for `df_expresion_data`

In [4]:
import numpy as np
import pandas as pd

# Numeric summary (column-wise) - EXPRESSION DATA
print("Numeric summary - Expression Data")
numeric_cols = df_expression_data.select_dtypes(include=[np.number]).columns.tolist()
numeric_summary = pd.DataFrame({
    "count": df_expression_data[numeric_cols].count(),
    "missing": df_expression_data[numeric_cols].isna().sum(),
    "missing_%": (df_expression_data[numeric_cols].isna().mean() * 100).round(2),
    "mean": df_expression_data[numeric_cols].mean(),
    "std": df_expression_data[numeric_cols].std(),
    "min": df_expression_data[numeric_cols].min(),
    "max": df_expression_data[numeric_cols].max(),
    "unique": df_expression_data[numeric_cols].nunique()
}).sort_values("missing_%", ascending=False)
print(numeric_summary)

Numeric summary - Expression Data
          count  missing  missing_%         mean          std  min      max  \
GSM36777  22283        0        0.0   913.726029  2732.391933  0.5  40917.1   
GSM36778  22283        0        0.0   973.443037  3371.973081  0.8  68495.3   
GSM36779  22283        0        0.0  1000.324961  3529.034928  0.3  64802.1   
GSM36780  22283        0        0.0  1034.967226  3891.188255  0.4  94763.8   
GSM36781  22283        0        0.0   985.429413  3552.996340  0.8  75816.4   
...         ...      ...        ...          ...          ...  ...      ...   
GSM37058  22283        0        0.0   973.229184  3230.747671  0.2  57830.1   
GSM37059  22283        0        0.0   966.519284  3220.683530  0.3  52801.8   
GSM37060  22283        0        0.0  1032.832074  3966.048197  0.3  70982.9   
GSM37061  22283        0        0.0   995.133075  3479.528686  0.9  65418.4   
GSM37062  22283        0        0.0   967.102800  3209.529465  0.4  51943.7   

          unique 

In [5]:
# Categorical summary (column-wise) - METADATA
print("\nCategorical summary - Metadata")
categorical_cols = df_metadata.select_dtypes(include=['object']).columns.tolist()

def safe_nunique(col):
    try:
        return col.astype(str).nunique()
    except:
        return 1

def safe_mode(col):
    try:
        return col.mode().iloc[0] if len(col.mode()) > 0 else 'N/A'
    except:
        return 'N/A'

def safe_top_freq(col):
    try:
        return col.astype(str).value_counts().iloc[0]
    except:
        return 0

cat_summary = pd.DataFrame({
    "count": df_metadata[categorical_cols].count(),
    "missing": df_metadata[categorical_cols].isna().sum(),
    "missing_%": (df_metadata[categorical_cols].isna().mean() * 100).round(2),
    "unique": [safe_nunique(df_metadata[col]) for col in categorical_cols],
    "top_value": [safe_mode(df_metadata[col]) for col in categorical_cols],
    "top_freq": [safe_top_freq(df_metadata[col]) for col in categorical_cols]
}, index=categorical_cols).sort_values("missing_%", ascending=False)

print(cat_summary)



Categorical summary - Metadata
                         count  missing  missing_%  unique  \
relation                   267       19       6.64       4   
geo_accession              286        0       0.00     286   
title                      286        0       0.00     286   
submission_date            286        0       0.00       1   
last_update_date           286        0       0.00       2   
type                       286        0       0.00       1   
channel_count              286        0       0.00       1   
source_name_ch1            286        0       0.00       1   
organism_ch1               286        0       0.00       1   
taxid_ch1                  286        0       0.00       1   
status                     286        0       0.00       1   
characteristics_ch1        286        0       0.00       2   
molecule_ch1               286        0       0.00       1   
platform_id                286        0       0.00       1   
description                286        

Expression matrix shows **full integrity** (0% missing, 22,283 probes × 286 samples) with wide dynamic range (0.2-94,763) and high variability (std ~3,000), ideal for clustering. Metadata is **99.3% complete** (`relation` only 6.64% missing/19 samples); critically, **survival data in `characteristics_ch1`** fully populated (2 patterns) alongside complete identifiers (`geo_accession`, `source_name_ch1="Breast"`).

## Target Leakage Prevention

To ensure valid prognostic stratification reflecting **baseline tumor biology only**, variables not plausibly available at the time of molecular profiling were excluded from clustering features prior to any preprocessing or feature engineering.

**Hard Leakage Variables (Reserved for Post-Hoc Analysis):** Outcome variables directly encoding relapse events were extracted from `characteristics_ch1` and **reserved exclusively** for survival analysis:

- `relapse:yes/no` event indicators
- `DFS_months` / time-to-relapse
- Distant metastasis status
- Post-treatment outcome annotations

These variables directly encode downstream survival outcomes and would invalidate unsupervised cluster discovery if used as clustering features.

**Soft Leakage Variables (Excluded from Clustering):** Clinically derived variables potentially contaminated by outcome knowledge were conservatively excluded:

- Any prognostic risk scores or indices from `characteristics_ch1`
- Post-profiling treatment response annotations  
- Derived clinical summaries dependent on follow-up data

Although some clinical descriptors (ER/PR status) may be partially available at baseline, their finalized annotations often incorporate longitudinal outcome data. Excluding them prevents reliance on downstream clinical summaries and ensures clusters reflect **pure pretreatment molecular heterogeneity**.

**Retained Features for Clustering:** Primary **gene expression matrix** (22,283 Affymetrix HG-U133A probes) + baseline sample descriptors only (`GSM_ID` for linkage). Survival endpoints reserved for Cox proportional hazards and Kaplan-Meier evaluation **after** molecular stratification.


In [ ]:
# Copy the original data to clean copies to preserve the original datasets
df_expression_data_clean = df_expression_data.copy()
df_metadata_clean = df_metadata.copy()

# Hard leakage: Features that should never be available at admission
hard_leakage_cols = [
    "Relapse Status",  # Example: this would be a post-relapse feature
    "Time to Relapse",  # Post-event information
    "Last Follow-Up",   # Information collected after treatment
    "Post-Treatment Response",  # Post-treatment info
    "Metastasis",  # Later-stage metastasis details
    "Disease Progression"  # Future progression after initial treatment
]

# Soft leakage: Typically determined after admission (conservative choice)
soft_leakage_cols = [
    "Tumor Stage",  # Could be determined post-admission, early stages onsidered
    "Survival Time",  # Time to event, should only be used for post-hoc analysis
    "Chemotherapy Type",  # Known after admission
    "Hormone Therapy Status",  # Typically determined after admission
    "Tumor Size",  # Might not be fully known at admission
]

# Remove hard leakage columns from df_expression_data (gene expression data)
df_expression_data_clean = df_expression_data_clean.drop(
    columns=[
        col for col in hard_leakage_cols if col in df_expression_data_clean.columns],
    errors="ignore"
)
# Remove **hard leakage** columns from df_metadata (metadata)
df_metadata_clean = df_metadata_clean.drop(
    columns=[
        col for col in hard_leakage_cols if col in df_metadata_clean.columns],
    errors="ignore"
)

# Remove soft leakage columns from df_expression_data (gene expression data)
df_expression_data_clean = df_expression_data_clean.drop(
    columns=[
        col for col in soft_leakage_cols if col in df_expression_data_clean.columns],
    errors="ignore"
)

# Remove **soft leakage** columns from df_metadata (metadata)
df_metadata_clean = df_metadata_clean.drop(
    columns=[
        col for col in soft_leakage_cols if col in df_metadata_clean.columns],
    errors="ignore"
)

# Check the cleaned datasets
print(f"Shape of df_expression_data after removing leakage columns: {
    df_expression_data_clean.shape}")
print(f"Shape of df_metadata after removing leakage columns: {
    df_metadata_clean.shape}")
df_expression_data_clean.head(), df_metadata_clean.head()


Shape of df_expression_data after removing leakage columns: (22283, 287)
Shape of df_metadata after removing leakage columns: (286, 27)


(      ID_REF  GSM36777  GSM36778  GSM36779  GSM36780  GSM36781  GSM36782  \
 0  1007_s_at    3848.1    6520.9    5285.7    4043.7    4263.6    2949.8   
 1    1053_at     228.9     112.5     178.4     398.7     417.7     221.2   
 2     117_at     213.1     189.8     269.7     312.4     327.1     225.0   
 3     121_at    1009.4    2083.3    1203.4    1104.4    1043.3    1117.6   
 4  1255_g_at      31.8     145.8      42.5     108.2      69.2      47.4   
 
    GSM36783  GSM36784  GSM36785  ...  GSM37053  GSM37054  GSM37055  GSM37056  \
 0    5498.9    3863.1    3370.4  ...    4058.2    4017.6    2841.0    2914.2   
 1     280.4     198.2     304.7  ...     183.4     356.1     234.6     169.4   
 2     243.5     244.4     348.5  ...     326.6     234.9     369.6     149.5   
 3    1085.4    1423.1    1196.4  ...    1041.3    1195.6     751.5    1117.8   
 4      84.3     102.0      22.8  ...     143.5      32.7      62.6      43.0   
 
    GSM37057  GSM37058  GSM37059  GSM37060  GSM3